File: src/main/java/ch/leadrian/samp/kamp/textkeygen/GenerateTextKeysTask.java
package ch.leadrian.samp.kamp.textkeygen;

import org.gradle.api.DefaultTask;
import org.gradle.api.tasks.InputFiles;
import org.gradle.api.tasks.OutputDirectories;
import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.List;
import java.util.Properties;
import java.util.Set;
import java.util.regex.Pattern;

import static java.util.Objects.requireNonNull;
import static java.util.regex.Pattern.compile;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toSet;

public class GenerateTextKeysTask extends DefaultTask {

    private static final Pattern STRINGS_FILE_PATTERN = compile("strings(_[a-z]{2}(_[A-Z]{2})?)?\\.properties");

    public GenerateTextKeysTask() {
        doLast(task -> generateTextKeys());
    }

    @InputFiles
    public List<File> getInputFiles() {
        TextKeysGeneratorPluginExtension extension = getExtension();
        File resourcesDirectoryFile = getFileFromObject(extension.getResourcesDirectory());
        return extension
                .getPackageNames()
                .stream()
                .flatMap(packageName -> getStringsPropertiesFiles(resourcesDirectoryFile, packageName).stream())
                .collect(toList());
    }

    @OutputDirectories
    public List<File> getOutputDirectories() {
        return getExtension().getPackageNames().stream().map(this::getOutputDirectory).collect(toList());
    }

    private void generateTextKeys() {
        TextKeysGeneratorPluginExtension extension = getExtension();
        File resourcesDirectoryFile = getFileFromObject(extension.getResourcesDirectory());
        extension.getPackageNames().forEach(packageName -> {
            List<File> stringsPropertiesFiles = getStringsPropertiesFiles(resourcesDirectoryFile, packageName);
            try {
                generateTextKeys(packageName, stringsPropertiesFiles);
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        });
    }

    private void generateTextKeys(String packageName, List<File> stringsPropertiesFiles) throws IOException {
        Set<String> stringPropertyNames = stringsPropertiesFiles
                .stream()
                .map(File::toPath)
                .map(this::loadProperties)
                .map(Properties::stringPropertyNames)
                .flatMap(Set::stream)
                .collect(toSet());
        TextKeysGenerator textKeysGenerator = new TextKeysGenerator();
        Path outputDirectory = getOutputDirectory(packageName).toPath();

        Files.createDirectories(outputDirectory);
        try (Writer writer = Files.newBufferedWriter(outputDirectory.resolve("TextKeys.java"), StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
            textKeysGenerator.generateTextKeyClasses("TextKeys", packageName, stringPropertyNames, writer);
        }
    }

    private List<File> getStringsPropertiesFiles(File resourcesDirectoryFile, String packageName) {
        String packagePath = packageNameToPath(packageName);
        try {
            return Files
                    .list(resourcesDirectoryFile.toPath().resolve(packagePath))
                    .filter(Files::isRegularFile)
                    .filter(path -> STRINGS_FILE_PATTERN.matcher(path.getFileName().toString()).matches())
                    .map(Path::toFile)
                    .collect(toList());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private Properties loadProperties(Path path) {
        Properties properties = new Properties();
        try (Reader reader = Files.newBufferedReader(path)) {
            properties.load(reader);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        return properties;
    }

    private File getFileFromObject(Object fileObject) {
        requireNonNull(fileObject);
        if (fileObject instanceof Path) {
            return ((Path) fileObject).toFile();
        } else if (fileObject instanceof File) {
            return (File) fileObject;
        } else {
            return new File(fileObject.toString());
        }
    }

    private TextKeysGeneratorPluginExtension getExtension() {
        return getProject().getExtensions().getByType(TextKeysGeneratorPluginExtension.class);
    }

    private File getOutputDirectory(String packageName) {
        return new File(getFileFromObject(getExtension().getOutputDirectory()), packageNameToPath(packageName));
    }

    @NotNull
    private String packageNameToPath(String packageName) {
        return packageName.replace('.', File.separatorChar);
    }

}

File: src/main/kotlin/ch/leadrian/samp/kamp/gradle/plugin/textkeygen/GenerateTextKeysTask.kt
package ch.leadrian.samp.kamp.gradle.plugin.textkeygen

import org.gradle.api.DefaultTask
import org.gradle.api.internal.file.FileLookup
import org.gradle.api.tasks.InputFiles
import org.gradle.api.tasks.OutputDirectories
import java.io.File
import java.io.IOException
import java.io.UncheckedIOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.util.*
import java.util.regex.Pattern.compile
import javax.inject.Inject
import kotlin.streams.toList

open class GenerateTextKeysTask
@Inject
constructor(private val fileLookup: FileLookup) : DefaultTask() {

    companion object {

        private val STRINGS_FILE_PATTERN = compile("strings(_[a-z]{2}(_[A-Z]{2})?)?\\.properties")
    }

    @InputFiles
    fun getInputFiles(): List<File> {
        val extension = extension
        val resourcesDirectoryFile = fileLookup.fileResolver.resolve(extension.resourcesDirectory)
        return extension.packageNames.flatMap { packageName -> getStringsPropertiesFiles(resourcesDirectoryFile, packageName) }
    }

    @OutputDirectories
    fun getOutputDirectories(): List<File> = extension.packageNames.map { this.getOutputDirectory(it) }

    private val extension: TextKeysGeneratorPluginExtension
        get() = project.extensions.getByType(TextKeysGeneratorPluginExtension::class.java)

    init {
        doLast { generateTextKeys() }
    }

    private fun generateTextKeys() {
        val extension = extension
        val resourcesDirectoryFile = fileLookup.fileResolver.resolve(extension.resourcesDirectory)
        extension.packageNames.forEach { packageName ->
            val stringsPropertiesFiles = getStringsPropertiesFiles(resourcesDirectoryFile, packageName)
            try {
                generateTextKeys(packageName, stringsPropertiesFiles)
            } catch (e: IOException) {
                throw UncheckedIOException(e)
            }
        }
    }

    @Throws(IOException::class)
    private fun generateTextKeys(packageName: String, stringsPropertiesFiles: List<File>) {
        val stringPropertyNames = stringsPropertiesFiles
                .map { it.toPath() }
                .map { this.loadProperties(it) }
                .flatMap { it.stringPropertyNames() }
                .toSet()
        val textKeysGenerator = TextKeysGenerator()
        val outputDirectory = getOutputDirectory(packageName).toPath()

        Files.createDirectories(outputDirectory)
        Files.newBufferedWriter(outputDirectory.resolve("TextKeys.java"), StandardOpenOption.CREATE, StandardOpenOption.WRITE).use { writer -> textKeysGenerator.generateTextKeyClasses("TextKeys", packageName, stringPropertyNames, writer) }
    }

    private fun getStringsPropertiesFiles(resourcesDirectoryFile: File, packageName: String): List<File> {
        val packagePath = packageNameToPath(packageName)
        try {
            return Files
                    .list(resourcesDirectoryFile.toPath().resolve(packagePath))
                    .filter { Files.isRegularFile(it) }
                    .filter { path -> STRINGS_FILE_PATTERN.matcher(path.fileName.toString()).matches() }
                    .map { it.toFile() }
                    .toList()
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }

    }

    private fun loadProperties(path: Path): Properties {
        val properties = Properties()
        try {
            Files.newBufferedReader(path, extension.charset).use { reader -> properties.load(reader) }
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }
        return properties
    }

    private fun getOutputDirectory(packageName: String): File {
        val outputDirectory = project.buildDir.resolve(TextKeysGeneratorPlugin.GENERATED_SOURCE_DIRECTORY)
        return File(outputDirectory, packageNameToPath(packageName))
    }

    private fun packageNameToPath(packageName: String): String = packageName.replace('.', File.separatorChar)

}
